File: LeetCode/982.minimum-increment-to-make-array-unique/minimum-increment-to-make-array-unique.cpp
class Solution {
public:
    int minIncrementForUnique(vector<int> &A) {
        int size = A.size();
        if (size == 0)  return 0;
        vector<int>tmp;
        map<int, int>mp;
        sort(A.begin(), A.end());
        int val = A[0];
        for (int i = 1; i < size; i++) {
            mp[A[i]] = 1;
            if (A[i] == val) {
                tmp.push_back(val);
            } else {
                val = A[i];
            }
        }
        size = tmp.size();
        int res = 0;
        if (size == 0) {
            res = 0;
        } else {
            int index = tmp[0];
            for (int i = 0; i < size; i++) {
                index = index < tmp[i] ? tmp[i] : index;
                while (mp[index] == 1 || index < tmp[i])  index++;
                mp[index] = 1;
                res += index - tmp[i];
            }
        }

        return res;
    }
};

File: LeetCode/0945.minimum-increment-to-make-array-unique/minimum-increment-to-make-array-unique.java
class Solution {
    public int minIncrementForUnique(int[] A) {
        if (A == null || A.length == 0) return 0;
        Arrays.sort(A);
        List<Integer>list = new ArrayList<>();
        Map<Integer,Integer>map = new HashMap<>();
        int val = A[0];
        for (int i = 1; i < A.length; i++){
            map.put(A[i], 1);
            if (A[i] == val)    list.add(val);
            else    val = A[i];
        }
        int res = 0;
        if (list.size() == 0)   res = 0;
        else{
            int index = list.get(0);
            for (int i = 0; i < list.size(); i++){
                val = list.get(i);
                index = index < val ? val : index;
                while (map.get(index) != null && map.get(index) == 1) index++;
                map.put(index, 1);
                res += index - val;
            }
        }
        return res;
    }
}